lib/stats.dx

```diff
@@ -0,0 +1,243 @@
+Distributions implement the `Dist` interface.
+
+interface Dist d a
+  log_pdf : d -> a -> Float     -- either the density function or mass function
+  log_cdf : d -> a -> Float     -- the cumulative distribution function (CDF)
+  log_sf : d -> a -> Float      -- the survivor function (complement of CDF)
+  quantile : d -> Float -> a    -- the quantile function (inverse CDF)
+  draw : d -> Key -> a          -- function for random draws
+
+
+'### Bernoulli distribution
+The [Bernoulli distribution](https://en.wikipedia.org/wiki/Bernoulli_distribution) is parameterised by its "success" probability, `p`.
+
+data BernoulliDist = Bernoulli Float
+
+instance Dist BernoulliDist Bool
+  log_pdf = \(Bernoulli p) b.
+    if (b)
+     then log p
+     else log1p (-p)
+  log_cdf = \(Bernoulli p) b.
+    if (b)
+      then 0.0
+      else log1p (-p)
+  log_sf = \(Bernoulli p) b.
+    if (b)
+      then -infinity
+      else log p
+  quantile = \(Bernoulli p) q.
+    q > (1 - p)
+  draw = \(Bernoulli p) k.
+    (rand k) < p
+
+
+'### Binomial distribution
+The [binomial distribution](https://en.wikipedia.org/wiki/Binomial_distribution) is parameterised by the number of trials, `n`, and the success probability, `p`. Mean is `np`.
+
+data BinomialDist = Binomial Nat Float
+
+def binomial_log_pdf_private (n: Nat) (p: Float) (x: Nat) : Float =
+  nf = n_to_f n
+  xf = n_to_f x
+  if (xf > nf)
+    then -infinity
+    else (xf * (log p)) + ((nf - xf) * (log1p (-p))) +
+      (lgamma (nf + 1)) - (lgamma (xf + 1)) - (lgamma (nf + 1 - xf))
+
+instance Dist BinomialDist Nat
+  log_pdf = \(Binomial n p) x.
+    binomial_log_pdf_private n p x
+  log_cdf = \(Binomial n p) x.
+    xp1:Nat = x + 1
+    logsumexp $ for i:(Fin xp1). binomial_log_pdf_private n p (ordinal i)
+  log_sf = \(Binomial n p) x.
+    nmx = n -| x
+    logsumexp $ for i:(Fin nmx). binomial_log_pdf_private n p (x + 1 + ordinal i)
+  quantile = \(Binomial n p) q.
+    np1:Nat = n + 1
+    lpdf = for i:(Fin np1). binomial_log_pdf_private n p (ordinal i)
+    cdf = cdf_for_categorical lpdf
+    mi = search_sorted cdf q
+    ordinal $ from_just mi
+  draw = \(Binomial n p) k.
+    sum $ map b_to_n (rand_vec n (draw $ Bernoulli p) k)
+
+-- TODO: remove "binom" from the prelude
+
+
+'### Exponential distribution
+The [exponential distribution](https://en.wikipedia.org/wiki/Exponential_distribution) is parameterised by its *rate*, `r > 0`, so that the mean of the distribution is `1/r`.
+
+data ExponentialDist = Exponential Float
+
+instance Dist ExponentialDist Float
+  log_pdf = \(Exponential r) x.
+    if (x < 0.0)
+      then -infinity
+      else (log r) - (r * x)
+  log_cdf = \(Exponential r) x.
+    if (x < 0.0)
+      then -infinity
+      else log1p (-exp (-r * x))
+  log_sf = \(Exponential r) x.
+    if (x < 0.0)
+      then 0.0
+      else -r * x
+  quantile = \(Exponential r) q.
+    (log1p (-q)) / -r
+  draw = \(Exponential r) k.
+    (log1p (-(rand k))) / -r
+
+
+'### Gaussian distribution
```

Comment:
This is minor, but I feel like Normal might be more standard? I guess we could support the Gaussian spelling through aliases
```
GaussianDist = NormalDist
Gaussian = Normal
```